# HiTEA
computational tool to identify trasposable element insertions using Hi-C data


- Description:

HiTEA (Hi-C based Transposable Element insertion Analysis) is geared to idenify non-reference transposable elemenet insertions from the Hi-C data. It uses split Hi-C read information and read coverage to detect insertions of three major classes of active human transposable elements (TE) - Alu (SINE), L1HS(LINE) and SVA.


- Dependancies:
  - PERL(≥v5.24)
  - R(≥v3.2)
  - bedtools(≥v2.26)
  - samtools(≥v1.7)
  - GNU-parallel [https://www.gnu.org/software/parallel/]
  - Pairtools [https://github.com/mirnylab/pairtools]
  - Following R packages are required for computation
    - GenomicRanges, data.table, MASS
  - Following optional R packages are required for HTML report generation
    - rmarkdown, knitr, EnrichedHeatmap, circlize
 
 - Required data files:
 HiTEA requires following reference data files during the run. These files are typically grouped under the directory with name of the genome build (such as hg19 or hg38). Following example illustrates the data directory format. 
   - hg38/indexC/        = The subdirectory contains a BWA index generated for the consensus TE sequences (consensus.fa)
   - hg38/RE/            = The subdirectory contains Juicer-formatted annotation of the RE sites along the genome (e.g. GATC.txt or AAGCTT.txt) 
   - hg38/indexP/        = This optional subdirectory contains polymorphic sequences for a given TE family. For example, all Polymorphic Alu sequences are written down in a file in Alu.fa file. BWA index is used for indexing these files. These sequences can also be retrieved for each sufamily from the Repbase database. For HiTEA run,this subfolder is not required.  
   - hg38/bgAnnotations/ = This subdirectory contains bed-formatted locations of the TE copies along the genome. For instance, All SINE copies are listed as Alu.bed or LINE copies are listed as L1.bed. The annotations can be easily downloaded from USCS table browser under repeats group [https://genome.ucsc.edu/] and segragated for the repeat classes that the user is interested in detecting.
   - hg38/repbase/       = The Repbase release enlist family-level consensus sequences for the TEs. They can be downloaded directly from the server and indexed using BWA index [https://www.girinst.org/repbase/]. 

 For generating Juicer-formatted RE annotations for enzymes other than MboI/DpnII/HindII/NcoI, user can use the script "generate_site_positions.py" available on Juicer Github repository [https://github.com/aidenlab/juicer] 
 
 
 
- Running HiTEA:
  *It is required to set the path-variable for HiTEA bash script file before running it.
  (At present, HiTEA requires that the data directory (i.e. directory containing reference files) to be present in the same directory where the scripts are located)  
 
 HiTEA can be run as a single line command. First, it checks for all dependancies and once satisfied, runs on the user-provided bam file. HiTEA accepts following input files
  - Lossless bam file: These name-sorted bam files can be downloaded from the 4DN data portal where the read-pairs carry classification tags generated by Pairtools [https://github.com/mirnylab/pairtools]
  - Alternately, HiTEA can use name sorted bam files as an input and employs Pairtools for read-type classification
  - If the user wants to use hg19 or hg38 genome assemblies with MboI/DpnII/HindIII/NcoI restriction endonuclease experiment, the precomputated files are available. User will need to input the genome build and RE motif information
  
```
$ hitea

****No input provided


Usage: hitea [-w workdir] [-p ncores] [-m enzyme] [-q anchor_mapq] [-o outprefix] [-s clip] [-g genome] [-r remap] [-x if WGS] [-h help] -i inputs (space separated psam/bam in inverted commas)


    -i inputs :          Input file in pairsam format or unsorted-lossless bam format
    -o outprefix :       Output prefix while generating report files (default: project)
    -m enzyme :          Restriction endunuclease used for the assay (default: '', available:MboI,DpnII,HindIII,Arima,NcoI,NotI)
    -w workdir:          Working directory where the files are to be written
    -p ncores :          Number of cores to be used (auto-detected)
    -q anchor_mapq :     Mapping quality threshold for repeat anchored mate (default: 28)
    -s clip :            Minimum clip length for detecting insertion (should be >=13bp) (default: 20)
    -g genome :          Genome build to be used (default:hg38, available: hg19)
    -r remap :           whether to remap unmapped clipped reads to the polymoprhic sequences (default:F)
    -x wgs :             whether the file is WGS experiment (default:F)
    -h help :            Display help message


```
  
Running HiTEA on a single input bam file
```
hitea -i bam/4DNFIC275NK8.bam -w GM12878_test -o gm12878 -g hg38 -m 'DpnII' -r 'T'
```

Running HiTEA on a list of input bam files of a single experiment
```
hitea -i 'bam/4DNFPC275NK8.bam bam/4DNFIJ275PQ9.bam bam/4DNFIC275HT2.bam' -w GM12878_test -o gm12878 -g hg38 -m 'DpnII' -r 'T'
```
  
